File: Data Structures & Algorithms/max-water-container/submission-5.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        maxArea,i,j=0,0,len(heights)-1
        while i<j:
            maxArea=max(maxArea,min(heights[i],heights[j])*(j-i))
            if heights[i]>heights[j]:
                j-=1
            else:
                i+=1
        return maxArea

File: Data Structures & Algorithms/eating-bananas/submission-5.js
class Solution {
    /**
     * @param {number[]} piles
     * @param {number} h
     * @return {number}
     */
    minEatingSpeed(piles, h) {
        let max=0
        for(const p of piles){
            max=Math.max(p,max)
        }
        if(piles.length===h){
            return max
        }
        let L=1,R=max,mid=0,hours=0,res=max
        while(L<=R){
            hours=0
            mid=Math.floor((L+R)/2)
            for(const p of piles){
                hours+=Math.ceil(p/mid)
                if(hours>h){
                    break
                }
            }
            if(hours<=h){
                res=Math.min(res,mid)
                R=mid-1
            }else{
                L=mid+1
            }            
        }
        return res
    }
}
